### PR TITLE
Fix UserAgentInfo and Constantpaths issues

### PR DESCRIPTION
**Pull request by kazrael2119**
fixes https://github.com/Azure/autorest.typescript/issues/2853
fixes https://github.com/Azure/autorest.typescript/issues/2852

### CODE REVIEW

File: packages/typespec-ts/src/modular/helpers/clientHelpers.ts [hunk 1]
```diff
@@ -287,17 +291,48 @@ function buildLoggingOptions(flavor?: PackageFlavor): string | undefined {
...
+    !clientPackageName.endsWith("-rest")
+  ) {
+    clientPackageName += "-rest";
+  }
+  const userAgentInfoStatement = sdkUserAgentPrefix.includes("api")
+    ? "const userAgentInfo = `azsdk-js-" +
+      clientPackageName +
+      "/" +
+      packageVersion +
+      "`;"
```
Comment (MaryGao) [inline packages/typespec-ts/src/modular/helpers/clientHelpers.ts:313]:
```suggestion
      "`;\n"
```

File: packages/typespec-test/test/loadtesting_modular/generated/typespec-ts/package.json [hunk 2]
```diff
@@ -94,6 +94,26 @@
     "test": "npm run clean && tshy && npm run unit-test:node && npm run unit-test:browser && npm run integration-test",
     "build": "npm run clean && tshy && npm run extract-api"
   },
+  "//metadata": {
```
Comment (MaryGao) [inline packages/typespec-test/test/loadtesting_modular/generated/typespec-ts/package.json:97]: Should not generate this in non-sdk repo.

File: packages/typespec-ts/src/modular/buildClientContext.ts [hunk 3]
```diff
@@ -21,6 +21,22 @@ import { getTypeExpression } from "./type-expressions/get-type-expression.js";
 import { resolveReference } from "../framework/reference.js";
 import { useDependencies } from "../framework/hooks/useDependencies.js";
 
+/**
+ * This function gets the path of the file containing the modular client context
+ */
+export function getContentPath(
```
Comment (MaryGao) [inline packages/typespec-ts/src/modular/buildClientContext.ts:27]:
```suggestion
export function getClientContextPath(
```

File: packages/rlc-common/src/metadata/buildPackageFile.ts [hunk 4]
```diff
@@ -57,6 +58,14 @@ export function buildPackageFile(
     packageInfo = buildAzureStandalonePackage(extendedConfig);
   }
 
+  if (
```
Comment (MaryGao) [inline packages/rlc-common/src/metadata/buildPackageFile.ts:61]: could you move this logic to buildAzureMonorepoPackage function?

File: packages/rlc-common/src/metadata/buildPackageFile.ts [hunk 5]
```diff
@@ -57,6 +58,14 @@ export function buildPackageFile(
...
+    model.options?.moduleKind == "esm" &&
```
Comment (MaryGao) [inline packages/rlc-common/src/metadata/buildPackageFile.ts:62]: why limit the moduleKind to esm? does that cjs would have no metadata?

Comment (kazrael2119) [reply packages/rlc-common/src/metadata/buildPackageFile.ts:62]: cjs will use the metadata from buildAzureMonorepoPackage  directly

File: packages/typespec-ts/src/modular/helpers/clientHelpers.ts [hunk 6]
```diff
@@ -287,18 +291,48 @@ function buildLoggingOptions(flavor?: PackageFlavor): string | undefined {
 
 export function buildUserAgentOptions(
   context: StatementedNode,
-  sdkUserAgentPrefix: string
+  sdkUserAgentPrefix: string,
+  codeModel: ModularCodeModel
```
Comment (MaryGao) [inline packages/typespec-ts/src/modular/helpers/clientHelpers.ts:295]: move codeModel as second param. this is convention.

File: packages/typespec-ts/src/modular/helpers/clientHelpers.ts [hunk 7]
```diff
@@ -287,18 +291,48 @@ function buildLoggingOptions(flavor?: PackageFlavor): string | undefined {
...
-    const prefixFromOptions = options?.userAgentOptions?.userAgentPrefix;
-    const userAgentPrefix = ${
-      "prefixFromOptions ? `${prefixFromOptions} " +
-      sdkUserAgentPrefix +
-      "` : " +
-      `"${sdkUserAgentPrefix}"`
-    };
-  `;
-
+  const userAgentStatementsArray = [];
```
Comment (MaryGao) [inline packages/typespec-ts/src/modular/helpers/clientHelpers.ts:297]:
```suggestion
  const userAgentStatements = [];
```

File: packages/typespec-ts/src/modular/helpers/clientHelpers.ts [hunk 8]
```diff
@@ -287,18 +291,48 @@ function buildLoggingOptions(flavor?: PackageFlavor): string | undefined {
...
+    "prefixFromOptions ? `${prefixFromOptions} " +
+    sdkUserAgentPrefix +
+    `${userAgentInfoStatement ? " ${userAgentInfo}" : ""}` +
+    "` : `" +
+    `${sdkUserAgentPrefix}` +
+    `${userAgentInfoStatement ? " ${userAgentInfo}`" : "`"}`
+  };`;
+  userAgentStatementsArray.push(userAgentPrefix);
+
+  const userAgentStatements = userAgentStatementsArray.join("\n");
```
Comment (MaryGao) [inline packages/typespec-ts/src/modular/helpers/clientHelpers.ts:335]: const userAgentStatements = userAgentStatementsArray.join("\n");
context.addStatements(userAgentStatements);

==>

context.addStatements( userAgentStatements.join("\n"));

File: packages/rlc-common/src/metadata/buildPackageFile.ts (hunk 4 above)
Comment (kazrael2119) [reply packages/rlc-common/src/metadata/buildPackageFile.ts:61]: this change should be an incremental update for typespec-ts.

File: packages/rlc-common/src/metadata/buildPackageFile.ts [hunk 9]
```diff
@@ -57,6 +58,14 @@ export function buildPackageFile(
...
+    model.options.isModularLibrary &&
```
Comment (MaryGao) [inline packages/rlc-common/src/metadata/buildPackageFile.ts:63]: do we have metadata in RLC?

Comment (kazrael2119) [reply packages/rlc-common/src/metadata/buildPackageFile.ts:63]: yes, if enable azureSdkForJs, metadata will be generated in typespec RLC, 
see https://github.com/Azure/autorest.typescript/pull/2888/files#diff-d58a1b40354ea90a08e1f21d4c6c55fd325197c24d6da40c2d96028a29fe4b16R474-R483

File: packages/typespec-ts/src/modular/helpers/clientHelpers.ts [hunk 10]
```diff
@@ -287,19 +291,48 @@ function buildLoggingOptions(flavor?: PackageFlavor): string | undefined {
...
+    "const prefixFromOptions = options?.userAgentOptions?.userAgentPrefix;";
+  userAgentStatements.push(prefixFromOptions);
+
+  let clientPackageName =
+    codeModel.options.packageDetails?.nameWithoutScope ??
+    codeModel.options.packageDetails?.name ??
+    "";
+  const packageVersion = codeModel.options.packageDetails?.version ?? "";
+  if (
+    !codeModel.options.isModularLibrary &&
```
Comment (MaryGao) [inline packages/typespec-ts/src/modular/helpers/clientHelpers.ts:308]: no need this.

File: packages/rlc-common/src/metadata/packageJson/buildAzureMonorepoPackage.ts [hunk 11]
```diff
@@ -20,9 +20,12 @@ export interface AzureMonorepoInfoConfig extends AzurePackageInfoConfig {
 /**
  * Builds the package.json for an Azure package that will be hosted in the azure-sdk-for-js mono repo.
  */
-export function buildAzureMonorepoPackage(config: AzureMonorepoInfoConfig) {
+export function buildAzureMonorepoPackage(
+  config: AzureMonorepoInfoConfig,
+  modularMetadata?: Record<string, any>
```
Comment (MaryGao) [inline packages/rlc-common/src/metadata/packageJson/buildAzureMonorepoPackage.ts:25]: move metadata in AzureMonorepoInfoConfig directly.

File: packages/rlc-common/src/metadata/packageJson/buildAzureMonorepoPackage.ts [hunk 12]
```diff
@@ -55,20 +58,9 @@ export function getAzureMonorepoDependencies(config: AzureMonorepoInfoConfig) {
  * Build the common package.json config for an Azure package that will be hosted in the azure-sdk-for-js mono repo.
  */
 export function getAzureMonorepoPackageInfo(
-  config: AzureMonorepoInfoConfig
+  config: AzureMonorepoInfoConfig,
+  modularMetadata?: Record<string, any>
```
Comment (MaryGao) [inline packages/rlc-common/src/metadata/packageJson/buildAzureMonorepoPackage.ts:62]: we don't need second param if we move into first one.

File: packages/rlc-common/src/metadata/packageJson/buildAzureMonorepoPackage.ts [hunk 13]
```diff
@@ -83,7 +75,9 @@ export function getAzureMonorepoPackageInfo(
       homepage: `https://github.com/Azure/azure-sdk-for-js/tree/main/${config.monorepoPackageDirectory}/README.md`
     }),
     prettier: "@azure/eslint-plugin-azure-sdk/prettier.json",
-    "//metadata": metadata
+    "//metadata": config.isModularLibrary
```
Comment (MaryGao) [inline packages/rlc-common/src/metadata/packageJson/buildAzureMonorepoPackage.ts:78]: "//metadata": config.isModularLibrary
      ? modularMetadata
      : getRLCMetadata(config)

==>

 "//metadata": config.isModularLibrary
      ? config.metadata
      : getRLCMetadata(config)

File: packages/rlc-common/test/integration/packageJson.spec.ts [hunk 14]
```diff
@@ -446,6 +446,41 @@ describe("Package file generation", () => {
         'dev-tool run vendored prettier --write --config ../../../.prettierrc.json --ignore-path ../../../.prettierignore "src/**/*.{ts,cts,mts}" "test/**/*.{ts,cts,mts}" "*.{js,cjs,mjs,json}" '
       );
     });
+
+    it("[esm] should have constantPaths in typespec modular", () => {
```
Comment (MaryGao) [inline packages/rlc-common/test/integration/packageJson.spec.ts:450]: should read metadata from config for modular

**Review by MaryGao — APPROVED** [review]

**Pull request merged**